File: internal/discovery/discovery.go
package discovery

import (
	"errors"
	"fmt"
	"sort"
	"sync"
	"time"

	"github.com/Georgecane/openhost/internal/capability"
	"github.com/Georgecane/openhost/internal/identity"
)

var (
	ErrInvalidAdvertisement = errors.New("invalid discovery advertisement")
	ErrParticipantExists    = errors.New("participant already discovered")
	ErrParticipantNotFound  = errors.New("participant not discovered")
	ErrInvalidParticipant   = errors.New("discovery participant identity must be valid")
	ErrInvalidTimestamp     = errors.New("discovery timestamp must not be zero")
	ErrInvalidFreshness     = errors.New("discovery freshness policy must be positive")
)

type Advertisement struct {
	Participant identity.Identity
	Capability  capability.Capability
	Sequence    uint64
	ObservedAt  time.Time
}

func (a Advertisement) Validate() error {
	if err := a.Participant.Validate(); err != nil || a.Participant.Kind != identity.ParticipantKind {
		return fmt.Errorf("%w: %w", ErrInvalidParticipant, err)
	}
	if err := a.Capability.Validate(); err != nil {
		return fmt.Errorf("%w: %w", ErrInvalidAdvertisement, err)
	}
	if a.ObservedAt.IsZero() {
		return fmt.Errorf("%w: %w", ErrInvalidAdvertisement, ErrInvalidTimestamp)
	}
	return nil
}

type State string

const (
	StateActive  State = "active"
	StateStale   State = "stale"
	StateExpired State = "expired"
)

type FreshnessPolicy struct {
	StaleAfter   time.Duration
	ExpireAfter  time.Duration
}

func (p FreshnessPolicy) Validate() error {
	if p.StaleAfter <= 0 || p.ExpireAfter <= 0 || p.StaleAfter >= p.ExpireAfter {
		return ErrInvalidFreshness
	}
	return nil
}

type Member struct {
	Advertisement Advertisement
	LastSeen      time.Time
}

func (m Member) Validate() error {
	if err := m.Advertisement.Validate(); err != nil {
		return err
	}
	if m.LastSeen.IsZero() {
		return fmt.Errorf("%w: %w", ErrInvalidAdvertisement, ErrInvalidTimestamp)
	}
	if m.LastSeen.Before(m.Advertisement.ObservedAt) {
		return fmt.Errorf("%w: last-seen timestamp precedes advertisement", ErrInvalidAdvertisement)
	}
	return nil
}

func (m Member) StateAt(now time.Time, policy FreshnessPolicy) (State, error) {
	if err := m.Validate(); err != nil {
		return "", err
	}
	if now.IsZero() {
		return "", ErrInvalidTimestamp
	}
	if err := policy.Validate(); err != nil {
		return "", err
	}

	age := now.Sub(m.LastSeen)
	if age < 0 {
		return "", fmt.Errorf("%w: state evaluated before last-seen timestamp", ErrInvalidTimestamp)
	}
	switch {
	case age >= policy.ExpireAfter:
		return StateExpired, nil
	case age >= policy.StaleAfter:
		return StateStale, nil
	default:
		return StateActive, nil
	}
}

type Registry interface {
	Upsert(advertisement Advertisement, observedAt time.Time) error
	Remove(participantID identity.Identity) error
	Get(participantID identity.Identity) (Member, error)
	Members() []Member
	MembersAt(now time.Time) ([]Member, error)
}

type MemoryRegistry struct {
	mu      sync.RWMutex
	members map[string]Member
	policy  FreshnessPolicy
}

var _ Registry = (*MemoryRegistry)(nil)

func NewMemoryRegistry(policy FreshnessPolicy) (*MemoryRegistry, error) {
	if err := policy.Validate(); err != nil {
		return nil, err
	}
	return &MemoryRegistry{
		members: make(map[string]Member),
		policy:  policy,
	}, nil
}

func (r *MemoryRegistry) Upsert(advertisement Advertisement, observedAt time.Time) error {
	if r == nil {
		return ErrInvalidAdvertisement
	}
	if err := advertisement.Validate(); err != nil {
		return err
	}
	if observedAt.IsZero() {
		return fmt.Errorf("%w: %w", ErrInvalidAdvertisement, ErrInvalidTimestamp)
	}
	if observedAt.Before(advertisement.ObservedAt) {
		return fmt.Errorf("%w: observation precedes advertisement", ErrInvalidAdvertisement)
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	key := advertisement.Participant.ID
	member, exists := r.members[key]
	if exists && advertisement.Sequence < member.Advertisement.Sequence {
		return fmt.Errorf("%w: sequence regressed", ErrInvalidAdvertisement)
	}

	r.members[key] = Member{
		Advertisement: advertisement,
		LastSeen:      observedAt,
	}
	return nil
}

func (r *MemoryRegistry) Remove(participantID identity.Identity) error {
	if r == nil {
		return ErrParticipantNotFound
	}
	if err := participantID.Validate(); err != nil || participantID.Kind != identity.ParticipantKind {
		return ErrInvalidParticipant
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if _, exists := r.members[participantID.ID]; !exists {
		return ErrParticipantNotFound
	}
	delete(r.members, participantID.ID)
	return nil
}

func (r *MemoryRegistry) Get(participantID identity.Identity) (Member, error) {
	if r == nil {
		return Member{}, ErrParticipantNotFound
	}
	if err := participantID.Validate(); err != nil || participantID.Kind != identity.ParticipantKind {
		return Member{}, ErrInvalidParticipant
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	member, exists := r.members[participantID.ID]
	if !exists {
		return Member{}, ErrParticipantNotFound
	}
	return member, nil
}

func (r *MemoryRegistry) Members() []Member {
	if r == nil {
		return nil
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	members := make([]Member, 0, len(r.members))
	for _, member := range r.members {
		members = append(members, member)
	}
	sort.Slice(members, func(i, j int) bool {
		return members[i].Advertisement.Participant.ID < members[j].Advertisement.Participant.ID
	})
	return members
}

func (r *MemoryRegistry) MembersAt(now time.Time) ([]Member, error) {
	members := r.Members()
	for _, member := range members {
		if _, err := member.StateAt(now, r.policy); err != nil {
			return nil, err
		}
	}
	return members, nil
}

func (r *MemoryRegistry) Policy() FreshnessPolicy {
	if r == nil {
		return FreshnessPolicy{}
	}
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.policy
}

File: src/discovery.rs
use crate::capability::Capability;
use crate::identity::{Identity, Kind};
use std::collections::BTreeMap;
use std::sync::RwLock;
use std::time::{Duration, SystemTime};

#[derive(Debug, Clone)]
pub struct Advertisement {
    pub participant: Identity,
    pub capability: Capability,
    pub sequence: u64,
    pub observed_at: SystemTime,
}

#[derive(Debug, Clone)]
pub struct Member { pub advertisement: Advertisement, pub last_seen: SystemTime }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum State { Active, Stale, Expired }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct FreshnessPolicy { pub stale_after: Duration, pub expire_after: Duration }

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum DiscoveryError { InvalidAdvertisement, Exists, NotFound, InvalidTimestamp, InvalidFreshness }

impl std::fmt::Display for DiscoveryError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { write!(f, "{self:?}") }
}
impl std::error::Error for DiscoveryError {}

impl FreshnessPolicy {
    pub fn validate(&self) -> Result<(), DiscoveryError> {
        if self.stale_after.is_zero() || self.expire_after.is_zero() || self.stale_after >= self.expire_after {
            return Err(DiscoveryError::InvalidFreshness);
        }
        Ok(())
    }
}

impl Advertisement {
    pub fn validate(&self) -> Result<(), DiscoveryError> {
        if self.participant.kind != Kind::Participant || self.participant.validate().is_err()
            || self.capability.validate().is_err() || self.observed_at == SystemTime::UNIX_EPOCH
        { return Err(DiscoveryError::InvalidAdvertisement); }
        Ok(())
    }
}

impl Member {
    pub fn state_at(&self, now: SystemTime, policy: FreshnessPolicy) -> Result<State, DiscoveryError> {
        if self.advertisement.validate().is_err() || self.last_seen == SystemTime::UNIX_EPOCH
            || self.last_seen < self.advertisement.observed_at || now < self.last_seen
        { return Err(DiscoveryError::InvalidTimestamp); }
        policy.validate()?;
        let age = now.duration_since(self.last_seen).map_err(|_| DiscoveryError::InvalidTimestamp)?;
        if age >= policy.expire_after { Ok(State::Expired) }
        else if age >= policy.stale_after { Ok(State::Stale) }
        else { Ok(State::Active) }
    }
}

pub trait Registry: Send + Sync {
    fn upsert(&self, advertisement: Advertisement, observed_at: SystemTime) -> Result<(), DiscoveryError>;
    fn remove(&self, id: &Identity) -> Result<(), DiscoveryError>;
    fn get(&self, id: &Identity) -> Result<Member, DiscoveryError>;
    fn members(&self) -> Vec<Member>;
    fn members_at(&self, now: SystemTime) -> Result<Vec<Member>, DiscoveryError>;
}

pub struct MemoryRegistry {
    members: RwLock<BTreeMap<String, Member>>,
    policy: FreshnessPolicy,
}

impl MemoryRegistry {
    pub fn new(policy: FreshnessPolicy) -> Result<Self, DiscoveryError> {
        policy.validate()?;
        Ok(Self { members: RwLock::new(BTreeMap::new()), policy })
    }

    pub fn policy(&self) -> FreshnessPolicy { self.policy }
}

impl Registry for MemoryRegistry {
    fn upsert(&self, advertisement: Advertisement, observed_at: SystemTime) -> Result<(), DiscoveryError> {
        advertisement.validate()?;
        if observed_at == SystemTime::UNIX_EPOCH || observed_at < advertisement.observed_at {
            return Err(DiscoveryError::InvalidTimestamp);
        }
        let mut members = self.members.write().expect("discovery lock poisoned");
        if let Some(existing) = members.get(&advertisement.participant.id) {
            if advertisement.sequence < existing.advertisement.sequence { return Err(DiscoveryError::InvalidAdvertisement); }
        }
        members.insert(advertisement.participant.id.clone(), Member { advertisement, last_seen: observed_at });
        Ok(())
    }

    fn remove(&self, id: &Identity) -> Result<(), DiscoveryError> {
        if id.kind != Kind::Participant || id.validate().is_err() { return Err(DiscoveryError::InvalidAdvertisement); }
        if self.members.write().expect("discovery lock poisoned").remove(&id.id).is_none() {
            return Err(DiscoveryError::NotFound);
        }
        Ok(())
    }

    fn get(&self, id: &Identity) -> Result<Member, DiscoveryError> {
        self.members.read().expect("discovery lock poisoned").get(&id.id).cloned().ok_or(DiscoveryError::NotFound)
    }

    fn members(&self) -> Vec<Member> {
        self.members.read().expect("discovery lock poisoned").values().cloned().collect()
    }

    fn members_at(&self, now: SystemTime) -> Result<Vec<Member>, DiscoveryError> {
        let members = self.members();
        for member in &members { member.state_at(now, self.policy)?; }
        Ok(members)
    }
}
